DS-2876 Allow categorical inputs to Jaccard and Correlation outputs

jrwishart opened the pull request:
- DS-2876 Remove redundant missing argument
- DS-2876 Coerce categorical variables to numeric
- DS-2876 Ensure sample sizes are computed correctly
- DS-2876 Handle categorical vars in correlation output
- DS-2876 Dichotomize outcome var in Jaccard Analysis
- DS-2876 Convert categorical to binary for Jaccard

jrwishart committed:
DS-2876 Remove redundant missing argument

The missing argument is no longer needed for Jaccard or Correlation
output types. The missing values are excluded or retained in the
estimation before the estimateImportance function is called.

jrwishart committed:
DS-2876 Coerce categorical variables to numeric

In the Correlation output type, categorical variables are coerced to
numeric using  flipTransformations::OrderedToNumeric. This is required
since other AsNumeric functions either convert to binary or indicator
functions for each level.

jrwishart committed:
DS-2876 Ensure sample sizes are computed correctly

jrwishart committed:
DS-2876 Handle categorical vars in correlation output

After the categorical variables have been converted to numeric in
previous commit. Variable names needed to be determined differently
since the estimation data will have indicator variables for each level.
Now the variable names are determined from the formula. This extraction
occurs after dummy variables are removed.

jrwishart committed:
DS-2876 Dichotomize outcome var in Jaccard Analysis

Use the dichotomizing outcome variable check for Jaccard coefficient
analysis as well. Also move the execution of code that checks data
suitable for Jaccard after this check. Code to be updated to check
predictors.

jrwishart committed:
DS-2876 Convert categorical to binary for Jaccard

Categorical predictors are coerced to binary by splitting each level
into a binary variable. Refactored the code was necessary since the
input.formula and other input arguments needed to allow for this
increase of predictor variables to analyse.

jrwishart committed:
DS-2876 Split ordinal var to binary num in Jaccard

Added code to ensure ordinal predictors are split into numeric binary
variables suitable for Jaccard coefficient analysis. Added more tests to
ensure this conversion occurs properly.

jrwishart committed:
DS-2876 Move broader Jaccard tests

Moved the broader Jaccard coefficient tests from test-dataproblems.R
to test-relativeimportance.R

mwmclean commented:
When I create a regression with a numeric binary predictor, `b`, and a categorical predictor, `c`, and outcome, `y`, with formula, `y~c+b`, `b` always appears before `c` in the results table when `output = "Jaccard Coefficient"`.

Should the reference/baseline level of a categorical predictor not be included in the results table since it is normally dropped in a Binary Logit regression?

jrwishart commented:
I can fix the ordering, would only need a more delicate construction of the data frame when the categorical variable is transformed to binary, reconciling it against the formula provided. 

The baseline level I didn't remove since the Jaccard coefficients are not the same as typical regression coefficients. Typical regression coefficients are fitted globally for all predictors in the model and have the first level as the baseline level of categorical predictors as part of the intercept (unless some other contrast choices are made of course). There the size of the categorical coefficients represent deviation from that baseline level. The Jaccard coefficient roughly represents pairwise overlap/agreement with each binary success between the outcome and individual predictor level. Even if a baseline level was chosen, omitting it from the table would mean the information of overlap between a baseline level successes and the outcome successes would be not shown and cant be deduced easily from the output.

jrwishart committed:
DS-2876 Ensure order for cat preds in Jaccard

Ensure the categorical variables in the data frame retain their correct
ordering and follow the provided formula to ensure the rows in the
output table are consistent with other outputs.
